### ISSUE TITLE
[Java backend] syntactic equality is getting stuck when term has existentials

### ISSUE THREAD

**Issue by osa1** [kframework/k-legacy#991]
I don't know much about backend parts but I'm hoping to fix this myself.

```
➜  issue696 git:(java8) ✗ cat test.k 
module TEST
     syntax KResult ::= Bool
     syntax KItem ::= "test"
     rule test => ?A:KResult ==K ?B:KResult
endmodule
```

Maude output:

```
<k>
    false
</k>
```

Java backend output:

```
<k>
    _35:KResult ==K _37:KResult
</k>
```

I think it's having problems with instantiating existentials. Or is this something else? @yilongli @andreistefanescu can you give me some tips?

**Comment by yilongli** [comment]
Unfortunately, `==K` is not really implemented as syntactic check in the Java backend. It's more like a semantic equality in some sense... I prefer to clearly separate them. @andreistefanescu should be able to tell you more about this.

**Comment by osa1** [comment]
@yilongli but doesn't that make programs run differently when different backends are used? In Maude, `==K` is syntactic, while in Java backend it's not, this should effect semantics.

**Comment by dwightguth** [comment]
Well, the maude support for symbolic stuff is deprecated and not really
intended to be used reasonably as a symbolic infrastructure, and for ground
terms, syntactic and semantic equality are the same.

**Comment by yilongli** [comment]
Theoretically speaking, I don't think syntactic and semantic equality are the same even for ground terms.

**Comment by osa1** [comment]
Is it possible to somehow alter semantics of `==K` ? How is it supposed to work for ground terms? Do we have a specification somewhere?

**Comment by yilongli** [comment]
`==K` is now implemented in `TermEquality#eq`. I think it should work as syntactic equality for ground terms.

**Comment by andreistefanescu** [comment]
Technically speaking, there is only one equality. There should be a distinction between normal variables and variables at the meta-level (`X = Y` should stay as it is, but `toMetaLevel(X) = toMetaLevel(Y)` should be false).
@osa1 this test behaves as intended in the java backend, and the maude backend provides limited support for variables. We should leave this one as it is.

**Comment by grosu** [comment]
Sorry for jumping into this discussion without checking all of it (ready to board an airplane in Toronto now), but the Maude backend is just wrong here.  We have decided that ==K is _mathematical equality_, so if it tells you true it means really true, if it tells you false it means really false, or otherwise should give you some equivalent form.

Grigore

**Comment by traiansf** [comment]
I disagree with @grosu .  ==K is (and always was) syntactic equality, and it is used all over the builtins specs, including in substitution where we need syntactic identity of variables.
I think =K was proposed as semantic equality but I'm not sure it was actually implemented.

**Comment by dlucanu** [comment]
I remember the same thing like Traian and we used in definitions the two equalities in this respect. I strongly believe that the two notations we agreed should be kept.

**Comment by grosu** [comment]
Yes, I think you are right, we decided to keep == for syntactic and = for semantic equality.

Grigore

**Comment by osa1** [comment]
OK, can we say that syntactic equality is decidable? I'm wondering about the correct behavior here. Currently `==K` in Java backend is getting stuck while evaluating `_35:KResult ==K _37:KResult`.

**Comment by dlucanu** [comment]
I guess that  `_35:KResult ==K _37:KResult` is rather a semantical equality than a syntactical one, i.e., it should be  `_35:KResult =K _37:KResult`. In that case it must be sent to the prover together with the path condition (we are doing something like that in the Maude-based symbolic backend).
I am not familiar with the Java backend, so this is just a guess.

**Comment by grosu** [comment]
Yes, it should be decidable.  We may consider using associativity and commutativity of certain constructs in the process, but syntactic equality should essentially be decidable (although, you can only trust it when it says "true"; "false" only means "I was not able to show them equal").

Grigore

**Comment by andreistefanescu** [comment]
Java backend considers `==K` to be the semantic equality because in most cases the user means semantics equality, which subsumes syntactic equality. Syntactic equality is not sufficient for symbolic execution. However, I decided agains changing `==K` to `=K` in the tutorial at that time, so this seemed as an acceptable compromise.
In the long run, I think we only need one equality, and differentiate between the equality between terms with variables (semantic equality) and equality between terms with variables represented at the meta-level (syntactic equality).

**Comment by osa1** [comment]
> We may consider using associativity and commutativity of certain constructs in the process, but syntactic equality should essentially be decidable.

But if we use some rules to decide for equality, doesn't that make it semantic equality instead of syntactic equality?

For example,

```
a + (b + c) ==K (a + b) + c
```

This should be false because syntactically two sides of the equation are not equal. If we apply associativity and rewrite this term to "true", then we make it semantic equality.

At least this is my understanding of equalities. (do they mean something different in this context?)

Basically I'd expect syntactic equality to be independent from any rules or definitions.

In this case original example should be rewritten to "false".

@andreistefanescu I'm confused, how can semantic equality work? Let's say LHS and RHS of the equation are not syntactically same. Now what? How can we apply rules to them to make them equal? (how to choose such rules?) Or how else can we decide for semantic equality?

**Comment by grosu** [comment]
@andreistefanescu It is a bad idea to have different semantics for ==K in the different backends.  It is OK to modify the tutorial.  I will do it.  I think we should fix this problem (i.e., modify both the Java backend and the Maude backend).  I'll add an issue about that.

@osa1 We tend to think about assoc and comm of operations as algorithmic properties, in that once you have a rewrite engine that optimizes for assoc and comm matching, terms which are equivalent modulo these operations are represented the same way internally.  That is, since + is assoc, the terms 
a + (b + c) and (a + b) + c are stored exactly the same way internally by the rewrite engine.  So it could actually be impossible to differentiate the two ...

Anyway, it does not really mater how much ==K does, as far as it includes syntactic equality and is fast.  Its main property is that t ==K t' implies t =K t', so people should use it cautiously, only in positive contexts.

**Comment by grosu** [comment]
@andreistefanescu actually, can you please add the issue yourself and assign me the task to modify the tutorial?
